### PR TITLE
fixed performance issue while resizing and resize event removal on destroy

### PR DESCRIPTION
**Pull request by rknoll**
1. fixes the performance issue while resizing (based on #860)
  - further references: #857, #939, #926, #857
2. fixes the `destroy` method when using multiple charts on a page
  - `window.onresize` got set to `null` so all other charts did not resize anymore

### CODE REVIEW

**Comment by aendrew** [comment]
Wait, this might fix #926? Somebody give this man a medal. :trophy: :+1: :shipit: 

Will evaluate and RTBC if it does as promised.

**Comment by rknoll** [comment]
:+1:
hmm.. I just realized, there is still a problem if we have to use `window.onresize`.. the events won't get released correctly if you don't call `destroy` in the reversed order in which you created the graphs :sweat:

I'll try to find a solution for that (at least for the case where c3 is the only code that uses `window.onresize` on the page), stay tuned..

**Comment by aendrew** [comment]
@rknoll Awesome, thanks for the update! 

Any idea whether it fixes how `generate()` keeps adding resize handlers when called on the same chart, leading to a massive memory hole over time? That's definitely something I want to test, but if it doesn't address that I won't bother. Thanks!

**Comment by rknoll** [comment]
Definitely, as long as you call `destroy()` on the old one at least..
Otherwise you'd have to check if there is already a Chart initialized for that DOM element and call its `destroy()` function. Maybe by checking if it has the `c3` class on it?

**Comment by jalebiman** [comment]
Hey all, thank you for this fix. Been struggling with this for about 2 weeks now and am very happy I stumbled upon this. This did indeed fix the leak when calling c3.generate repeatedly on the same chart. Thank you for the fix.

**Comment by ergo** [comment]
Wow this is awesome, I need this :D

**Comment by aendrew** [comment]
Here's the results of two benchmarks, first without calling `destroy()` between `generate()` calls and the latter with it:
no destroy calling

with destroy being called

[Plunker used](http://embed.plnkr.co/Xpu57IMAkDtxLiq9HXrN/preview)

Both show a steady rise in the number of listeners and ongoing memory leaks, though it seems like GC is working WAY better in the second instance. While it may not fix #926 entirely, it's definitely a step in the right direction.

---

That said, this isn't exactly what this PR is supposed to fix. It's supposed to be nullify the onresize events such that resizing the browser doesn't cause a memory leak (At least, if I'm understanding properly — @rknoll, please do let me know if I'm missing the point). The following graph shows me resizing the chart a bunch of times:

resizing

Maybe it's just because the buffer filled too quickly and all I've recorded is two resizes, but it seems like not a lot of spiking in the middle of the test, which is a good sign? Again, I'm really not the person who should be doing this, I really am not sure what I'm doing here... Will keep testing.

tl;dr — this PR seems to offer a pretty decent performance boost, and even if it doesn't resolve the aforementioned issues, it's definitely a step in the right direction and if nothing else gives us an internal API for continuing to attack resize-related issues. LGTM; :+1: for merge.

@rknoll That said, mind squashing your commits? Thanks!

**Comment by aendrew** [comment]
I still vote to merge it, but I would be forever grateful if somebody with more of a clue what they're doing re: JavaScript performance profiling could take a look at it.

**Comment by rknoll** [comment]
what I found is that if you manually trigger the GC in chrome, it will remove all listeners again.. I don't really have too much insights into the whole project to tell why it won't get freed by itself, and, I'm not a JS ninja either :-/
The main reason for this PR (at least for me) was that I have multiple C3 charts on a page, that get `destroy()`-ed and recreated individually, and the original implementation prevented me from destroying the graphs as it just removed the resize listeners completely (from all other graphs as well)..
Furthermore I really wanted to get #860 in as well to reduce the load on pages with more graphs or on mobile..
So even though it does not fix the "leaks" (without manually triggering GC), I'd also like to see this in the next release, at least for the 2 mentioned reasons :-) That said, I'll squash right now ;-)

**Comment by masayuki0812** [comment]
I merged this PR. Thanks a lot! @rknoll 
One thing. I removed `resize.timeout` option because it's a kind of internal things and `100` seems enough for any case basically.
